src/_recursiveMerge.ts

```diff
@@ -0,0 +1,44 @@
+import { uniq } from './uniq';
+import { isCyclic } from './_isCyclic';
+
+export type DeepPartialArray<T> = ReadonlyArray<DeepPartial<T>>;
+// eslint-disable-next-line @typescript-eslint/ban-types
+export type DeepPartial<T> = T extends Function
+  ? T
+  : T extends ReadonlyArray<infer U>
+  ? DeepPartialArray<U>
+  : T extends object
+  ? DeepPartialObject<T>
+  : unknown;
+export type DeepPartialObject<T> = { [P in keyof T]?: DeepPartial<T[P]> };
+
+/**
+ * Recursively merges two values, `a` and `b`.
+ *
+ * @param {unknown} a - The first value to merge. This is the dominant parameter in the merge operation.
+ * @param {unknown} b - The second value to merge.
+ * @returns {unknown} The result of the merge operation.
+ */
+export function recursiveMerge<T>(a: T, b: unknown): T {
+  if (
+    typeof a !== 'object' ||
+    typeof b !== 'object' ||
+    a === null ||
+    b === null
+  ) {
+    return a ?? (b as T);
+  }
+  if (Array.isArray(a)) {
+    if (Array.isArray(b)) {
+      return uniq(a.concat(b)) as T;
+    }
+    return a;
+  }
+  const output = { ...a };
+  const keys = uniq(Object.keys(a).concat(Object.keys(b)));
+  for (const k of keys) {
+    // @ts-expect-error We've already ensured that "a" and "b" are objects at runtime.
+    if (!isCyclic(a[k])) output[k] = recursiveMerge(a[k], b[k]);
+  }
```

Comment:
I think you are over-complicating what you need to do here, this is also the reason why the typing becomes harder. The type for recursiveMerge should be:
```ts
recursiveMerge(a: object, b: object): T {
```

This ensures that you are only dealing with merging for objects, anything else (`undefined`, strings, numbers, booleans, etc...) simply won't be callable.

Now, you simply do the merge of objects like this:
```suggestion
  // At this point the output is already merged, simply not deeply merged.
  const output = { ...a, ...b };
  // now just scan the output and look for values that should have been deep-merged
  for (const k in b) {
    const aValue = a[k];
    const bValue = b[k];
    if (
      typeof aValue === "object" &&
      !isCyclic(a[k]) &&
      typeof bValue === "object"
    ) {
      // These are the only keys that need recursive merging. At this point
      // we already know that both of them are objects, so they match the type
      // of recursiveMerge.
      output[k] = resursiveMerge(aValue, bValue);
    }
  }
```

Reply:
done